uxarray/grid/neighbors.py

```diff
@@ -0,0 +1,220 @@
+
+from uxarray.constants import INT_DTYPE
+
+
+class BallTree:
+    """Custom BallTree datastructure written around the
+    ``sklearn.neighbors.BallTree`` implementation for use with either the
+    corner (``Mesh2_node_x``, ``Mesh2_node_y``) or center (``Mesh2_face_x``,
+    ``Mesh2_face_y``) nodes of the inputted unstructured grid.
+
+    Notes
+    -----
+    See `sklearn.neighbors.BallTree <https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.BallTree.html>`__
+    for further information about the wrapped data structure.
+    """
+
+    def __init__(self, grid, node_type='corner'):
+
+        # construct tree from corner nodes
+        if node_type == "corner":
+
+            XY = np.vstack((deg2rad(grid.Mesh2_node_y.values),
+                            deg2rad(grid.Mesh2_node_x.values))).T
+
+            self.n_elements = grid.nMesh2_node
+            self.tree = SKBallTree(XY, metric='haversine')
+
+        # construct tree from center nodes
+        elif node_type == "center":
+
+            # warning until we work on how face centers are represented and constructed
+            warnings.warn(
+                "Internal representation of face center coordinates has not been tested or verified. Results"
+                "may not be as expected until this has been patched in the Grid class."
+            )
+
+            # center nodes must be present, remove once construction methods for them is added
+            if grid.Mesh2_face_x is None or grid.Mesh2_face_y is None:
+                raise ValueError
+
+            XY = np.vstack((deg2rad(grid.Mesh2_face_y.values),
+                            deg2rad(grid.Mesh2_face_x.values))).T
+
+            self.n_elements = grid.nMesh2_face
+            self.tree = SKBallTree(XY, metric='haversine')
+
+    def query(self,
+              xy: Union[np.ndarray, list, tuple],
+              k: Optional[int] = 1,
+              use_radians: Optional[bool] = False,
+              return_distance: Optional[bool] = True,
+              dualtree: Optional[bool] = False,
+              breadth_first: Optional[bool] = False,
+              sort_results: Optional[bool] = True):
+        """Query the tree for the ``k`` nearest neighbors.
+
+        Parameters
+        ----------
+        xy : array_like
+            coordinate pairs in degrees (lon, lat) to query
+        k: int, default=1
+            The number of nearest neighbors to return
+        use_radians : bool, optional
+            if True, queries assuming xy are inputted in radians, not degrees
+        return_distance : bool, optional
+            if True, return a tuple ``(d, i)`` of distances and indices if False, return array i
+        dualtree : bool, default=False
+            if True, use the dual tree formalism for the query: a tree is built for the query points, and the pair of trees is used to efficiently search this space. This can lead to better performance as the number of points grows large.
```

Comment:
Text too long for a single line. Curious how it passed the pre-commit; did we change the pre-commit configs/dependencies recently?

Reply:
I was surprised by this also, not sure why it was allowed.